ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,530 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief This file implements the netevent event-monitor program type hook on eBPF for Windows.
+ */
+
+#include "ebpf_netevent_hooks.h"
+#include "netevent_ebpf_ext_event.h"
+#include "netevent_ebpf_ext_program_info.h"
+
+#include <errno.h>
+
+//
+// Global variables.
+//
+
+// Define a dynamic event buffer for optimizing the event data copy.
+static uint8_t* _event_buffer = NULL; ///< Event buffer for copying the event data.
+static size_t _event_buffer_size =
+    4096; ///< Initial size of the event buffer, which will be dynamically resized as needed.
+EX_PUSH_LOCK _ebpf_netevent_push_event_lock;
+
+// Define the GUID for the NetEvent NPI (must match the one of the provider)
+const NPIID netevent_npiid = {0x2227e819, 0x8d8b, 0x11d4, {0xab, 0xad, 0x00, 0x90, 0x27, 0x71, 0x9e, 0x09}};
+// Define the client module's ID
+const NPI_MODULEID netevent_client_module_id = {
+    sizeof(NPI_MODULEID), MIT_GUID, {0x8a9a5ef1, 0x2aa1, 0x42e9, {0x89, 0x5, 0xd1, 0xcf, 0x6, 0xc5, 0x77, 0x64}}};
+
+//
+// Prototypes.
+//
+static ebpf_result_t
+_ebpf_netevent_program_context_create(
+    _In_reads_bytes_opt_(data_size_in) const uint8_t* data_in,
+    size_t data_size_in,
+    _In_reads_bytes_opt_(context_size_in) const uint8_t* context_in,
+    size_t context_size_in,
+    _Outptr_ void** context);
+
+static void
+_ebpf_netevent_program_context_destroy(
+    _In_opt_ void* context,
+    _Out_writes_bytes_to_(*data_size_out, *data_size_out) uint8_t* data_out,
+    _Inout_ size_t* data_size_out,
+    _Out_writes_bytes_to_(*context_size_out, *context_size_out) uint8_t* context_out,
+    _Inout_ size_t* context_size_out);
+
+static void
+_ebpf_netevent_push_event(_In_ netevent_event_md_t* netevent_event);
+
+NTSTATUS
+_netevent_ebpf_extension_attach_provider(
+    _In_ HANDLE nmr_binding_handle,
+    _In_ void* client_context,
+    _In_ PNPI_REGISTRATION_INSTANCE provider_registration_instance);
+
+NTSTATUS
+_netevent_ebpf_extension_detach_provider(_In_ HANDLE nmr_binding_handle);
+
+//
+// Structures for attaching to NetEvent (as an NMR client)
+//
+
+// Dispatch table for the client module's helper functions
+static const void* _ebpf_netevent_ext_helper_functions[] = {(void*)&_ebpf_netevent_push_event};
+const ebpf_helper_function_addresses_t _netevent_client_dispatch = {
```

Comment:
@gtrevi - this is the contract between netevent NPI provider (pktmon) and netevent NPI client (ebpf extension). This must not use `ebpf_helper_function_addresses_t` type. Instead it should use an array of function pointers such as:
```
typedef NTSTATUS (*netevent_push_event_t)(
    _In_ const netevent_event_md_t* netevent_event);
    
typedef struct _net_event_npi_dispatch_table
{
    uint16_t version; ///< Version of the dispatch table.
    uint16_t count;   ///< Number of entries in the dispatch table.
    netevent_push_event_t push_event;
    ...
} net_event_npi_dispatch_table_t;
```

Reply:
@shankarseal I guess you're referring to the `netevent_sim` as the provider.
Can you please elaborate on why the library's pre-defined dispatch table `ebpf_helper_function_addresses_t` is not suitable for being used? 
It has a versioning header, a function count and an array of function pointers, exactly what is needed (and what in the custom type you're suggesting). 

```c
typedef struct _ebpf_helper_function_addresses
{
    ebpf_extension_header_t header;
    uint32_t helper_function_count;
    uint64_t* helper_function_address;
} ebpf_helper_function_addresses_t;
```

Is there any reference to scope usage that I can refer to?

Reply:
closing as agreed